### ISSUE TITLE
Transfer Oefen dataset to NGT dataset?

### ISSUE THREAD

**Issue by uklomp** [Signbank/Global-signbank#1119]
Would it be possible to put all our entries from the Oefen dataset (all the lemma's, glosses, video's and additional descriptive information) into the NGT dataset? 
I dont think that you will encounter the problem that some of these entries already exist in the NGT dataset, but if you do, the original entry in the NGT dataset can stay (and the Oefen one can be skipped), or the NGT one can be overwritten, both is fine.

**Comment by susanodd** [comment]
i will look into how best to do this. This needs to be done with a command rather than export/import.

The IDs of the glosses and lemmas are already in use (by themselves), so it should be possible to just change the dataset and keep the same IDs. It would be best to move them rather than copy them otherwise they don't keep the same IDs.

The media files would also need to be moved by the command. But there are links to these files inside of the GlossVideo class. (It keeps track of video operations and makes backups.)

**Comment by susanodd** [comment]
Would you need commands over datasets? (Say like queries to find intersections or exclusion over two datasets?)

I'm trying to figure out what you would want to do with this command if it needs to be in the interface rather than a straightforward command.

We have Copy at the gloss level and you can copy the gloss from one dataset to another (Both must be selected and you need permissions in both.) But this command only copies Phonology, not senses or relations.

**Comment by susanodd** [comment]
@uklomp If you just want to create the same signs in dataset NGT, then this can be done via CSV.
The signs will all get new Signbank IDs.
The creation creates new lemmas at the same time as new glosses.
For creation, just the Dataset, and the Lemma and Annotation columns.

The after creating new signs via CSV in NGT, you need to export those to obtain the new Signbank ID in the new dataset. Then you can update the new signs.
From your original export of those signs from Oefen, you can copy the columns you want to update (the other fields).

As you can guess, this would be error-prone, since you need to use the IDs of the new glosses for the right signs. I guess if you sort your original file by the ID, then it will create the new glosses in the order they appear.
So you could sort the "update" of those glosses the same way.

Once you have created the new signs in NGT (from your signs that are in Oefen), you can search the NGT glosses on Created By and Creation Date to limit the results to those new signs. Then export the new signs to CSV.

They should be sorted in the same order as the original signs, so you could copy the phonology and senses columns to the new file to make it ready for import.

**Comment by susanodd** [comment]
@uklomp Are you able to do the above yourself, or would you rather have me do it?
It needs two steps. Two CSV files for import into NGT.
The first needs to done first to create the new signs and get new IDs.

If you want to "move" the glosses into NGT, then this needs to be implemented as a command instead of CSV.
Or you need to use the Copy button in Gloss Detail View for each sign. (And have both datasets selected to copy from one to the other.)

**Comment by susanodd** [comment]
Doing the videos, @Jetske would know best, since they need to be copied into a different dataset directory.
This is more tricky, because of how they are stored.

**Comment by uklomp** [comment]
Hi Susan, I would rather have you do it, if you have time and space. I’ll look into all your messages again on Tuesday. So if things are not clear in my request, I can specify it again on Tuesday. There is no hurry 😊
To get back to one question here is: it is indeed about moving all the content of Oefen into the NGT dataset.

**Comment by susanodd** [comment]
Okay, as per the caps csv issue, I modified those recent glosses that were not in caps.
I went ahead and removed the new ones that were duplicates.

I just browsed the Oefen dataset. One thing that we probably want to do is to allow multiple videos for glosses. @Jetske and I talked about this before. There is a Video View for glosses (panel you can click on) that also shows the other media, including other videos users have uploaded for the gloss.
Since your Oefen dataset also contains different videos for signs that already exist, these should be put in the other media or the original (old) videos could be put there. That is a task for @Jetske to figure out with the videos.

**Comment by susanodd** [comment]
I made a Move button in Gloss Detail Edit View.
This is fairly straightforward and it works!

The permissions for this button to show up and be used will probably need to be super user.

I made a signal function for gloss videos so that when the lemma (dataset field of LemmaIdgloss) is updated, the function is invoked and the video should be moved. (There are quite a few of these functions.)
I need to confirm that it works.

**Comment by susanodd** [comment]
It works it works! The code I wrote earlier this afternoon and tried moving a sign. After getting bored with Disney, I just tried it out with a video.

It works! It moves the videos to the right location, too.

I only need to add some permission checks on the button. 
And some checks that the dataset languages are the same for the target dataset. (I have no idea what would happen if there are other languages. I don't think it would fail they'd just become hidden.)

**Comment by susanodd** [comment]
I made it so the new Move button is only for superusers.
I added code to also move the related glosses, including morphology (Sequential, Blend).
The command successfully moves the gloss and the component glosses to the target dataset.
The videos and images are moved as well!
Here is a screenshot the glosses of tstMH as the only dataset chosen.
You can see the moved glosses here. This is on Safari.

![gloss-ijsbeer-moved-to-tstMH](https://github.com/Signbank/Global-signbank/assets/24272544/0da44d04-6772-448a-ae7b-f79737a65d13)

**Comment by susanodd** [comment]
![move-robin](https://github.com/Signbank/Global-signbank/assets/24272544/19595e1f-12d3-4967-a57b-e900e0333b3c)

**Comment by susanodd** [comment]
Here's the gloss in NGT

![robin-relations-ngt](https://github.com/Signbank/Global-signbank/assets/24272544/452b6ed5-6968-4983-bd1d-47e750ace2bc)

**Comment by susanodd** [comment]
And the relations view of ijsbeer (after it was moved to tstMH)

![ijsbeer-relations](https://github.com/Signbank/Global-signbank/assets/24272544/cf05f906-0980-4a64-87e4-7be8fd9944c9)

**Comment by susanodd** [comment]
@Jetske, signbank-dev is live again. You can demo the Move button.

**Comment by uklomp** [comment]
great, I tried this and it works! 
Is it also possible to move the whole set or do I need to move the signs individually?

**Comment by susanodd** [comment]
@uklomp you need to move them individually.
I thought it would be safer that way, since the command also looks at related glosses.
Those are moved too. It tells you what the related glosses are.

**Comment by susanodd** [comment]
This is completed.
I needed to remove some lemma's from NGT that had no glosses associated with them that overlapped with the animals in Oefen.
I also needed to add a -A suffix to a couple of glosses that overlapped.

To make it simpler to move, I added a Semantics Animal indicator for the glosses, to allow searching on Animals.
Otherwise it was impossible to find the Oefen glosses mixed in with the four thousand NGT glosses.
